File: packages/python/src/teleglance/parsing/channel.py
"""Parsers for channel-level information and page-status detection.

Like the message parsers, all structure comes from :class:`Selectors` —
nothing here hard-codes t.me class names.
"""

from __future__ import annotations

from selectolax.lexbor import LexborHTMLParser, LexborNode

from .._utils import clean_text, parse_count
from ..models import Channel, ChannelCounts
from .entities import inner_html
from .selectors import DEFAULT_SELECTORS, Selectors


class PageKind:
    """What a fetched t.me page turned out to be."""

    FEED = "feed"  # /s/ preview with channel info (may have zero messages)
    CARD = "card"  # plain t.me/<name> profile card (no public preview from /s/)
    NOT_FOUND = "not_found"
    UNKNOWN = "unknown"  # neither — likely markup drift or an interception page


def classify_page(html: str, selectors: Selectors | None = None) -> str:
    sel = selectors or DEFAULT_SELECTORS
    tree = LexborHTMLParser(html)
    if tree.css_first(sel.channel_info) is not None:
        return PageKind.FEED
    if tree.css_first(sel.card_title) is not None:
        return PageKind.CARD
    if tree.css_first(sel.not_found) is not None:
        return PageKind.NOT_FOUND
    return PageKind.UNKNOWN


def _text_of(container: LexborNode, selector: str) -> str | None:
    node = container.css_first(selector)
    return clean_text(node.text(deep=True)) if node is not None else None


def _html_of(container: LexborNode, selector: str) -> str | None:
    node = container.css_first(selector)
    return inner_html(node) if node is not None else None


def parse_channel(html: str, username: str, selectors: Selectors | None = None) -> Channel | None:
    """Parse channel info from either a /s/ feed page (rich: counters) or a
    profile card page (title/description only). Returns None if the page has
    no channel data at all."""
    sel = selectors or DEFAULT_SELECTORS
    tree = LexborHTMLParser(html)

    info = tree.css_first(sel.channel_info)
    if info is not None:
        username_node = info.css_first(sel.channel_username)
        if username_node is not None:
            handle = (clean_text(username_node.text()) or "").lstrip("@") or username
        else:
            handle = username

        raw_counts: dict[str, str] = {}
        for counter in info.css(sel.channel_counter):
            value = _text_of(counter, sel.counter_value)
            kind = _text_of(counter, sel.counter_type)
            if value and kind:
                raw_counts[kind] = value

        def count_of(*labels: str) -> int | None:
            for label in labels:
                if label in raw_counts:
                    return parse_count(raw_counts[label])
            return None

        avatar = info.css_first(sel.channel_avatar)
        return Channel(
            username=handle,
            url=f"https://t.me/{handle}",
            title=_text_of(info, sel.channel_title) or handle,
            description=_text_of(info, sel.channel_description),
            description_html=_html_of(info, sel.channel_description),
            avatar_url=avatar.attributes.get("src") if avatar is not None else None,
            counts=ChannelCounts(
                subscribers=count_of("subscribers", "subscriber"),
                photos=count_of("photos", "photo"),
                videos=count_of("videos", "video"),
                files=count_of("files", "file"),
                links=count_of("links", "link"),
                raw=raw_counts,
            ),
        )

    title = _text_of(tree.root, sel.card_title) if tree.root is not None else None
    if title is not None:
        avatar = tree.css_first(sel.card_avatar)
        raw_counts = {}
        subscribers = None
        extra_text = _text_of(tree.root, sel.card_extra)
        if extra_text and "subscriber" in extra_text:
            raw_counts["subscribers"] = extra_text
            subscribers = parse_count(extra_text)
        return Channel(
            username=username,
            url=f"https://t.me/{username}",
            title=title,
            description=_text_of(tree.root, sel.card_description),
            description_html=_html_of(tree.root, sel.card_description),
            avatar_url=avatar.attributes.get("src") if avatar is not None else None,
            counts=ChannelCounts(subscribers=subscribers, raw=raw_counts),
        )

    return None

File: packages/ts/src/parsing/channel.ts
/**
 * Parsers for channel-level information and page-status detection.
 *
 * Like the message parsers, all structure comes from Selectors —
 * nothing here hard-codes t.me class names.
 */

import { parse } from 'node-html-parser';
import type { HTMLElement } from 'node-html-parser';
import { cleanText, parseCount } from '../utils.js';
import type { Channel, ChannelCounts } from '../models/channel.js';
import { DEFAULT_SELECTORS, type Selectors } from './selectors.js';

/**
 * What a fetched t.me page turned out to be.
 */
export enum PageKind {
  FEED = 'feed', // /s/ preview with channel info (may have zero messages)
  CARD = 'card', // plain t.me/<name> profile card (no public preview from /s/)
  NOT_FOUND = 'not_found',
  UNKNOWN = 'unknown', // neither — likely markup drift or an interception page
}

/**
 * Classify a t.me page by its structure.
 */
export function classifyPage(
  html: string,
  selectors: Selectors = DEFAULT_SELECTORS,
): PageKind {
  const tree = parse(html);

  if (tree.querySelector(selectors.channelInfo)) {
    return PageKind.FEED;
  }
  if (tree.querySelector(selectors.cardTitle)) {
    return PageKind.CARD;
  }
  if (tree.querySelector(selectors.notFound)) {
    return PageKind.NOT_FOUND;
  }
  return PageKind.UNKNOWN;
}

function textOf(container: HTMLElement, selector: string): string | null {
  const node = container.querySelector(selector);
  return node ? cleanText(node.textContent) : null;
}

function htmlOf(container: HTMLElement, selector: string): string | null {
  const node = container.querySelector(selector);
  return node ? node.innerHTML : null;
}

/**
 * Parse channel info from either a /s/ feed page (rich: counters) or a
 * profile card page (title/description only). Returns null if the page has
 * no channel data at all.
 */
export function parseChannel(
  html: string,
  username: string,
  selectors: Selectors = DEFAULT_SELECTORS,
): Channel | null {
  const tree = parse(html);

  // Try feed page first (rich data with counters)
  const info = tree.querySelector(selectors.channelInfo);
  if (info) {
    const usernameNode = info.querySelector(selectors.channelUsername);
    const handle = usernameNode
      ? (cleanText(usernameNode.textContent) || '').replace(/^@/, '') || username
      : username;

    const rawCounts: Record<string, string> = {};
    const counterNodes = info.querySelectorAll(selectors.channelCounter);
    for (const counter of counterNodes) {
      const value = textOf(counter, selectors.counterValue);
      const kind = textOf(counter, selectors.counterType);
      if (value && kind) {
        rawCounts[kind] = value;
      }
    }

    const countOf = (...labels: string[]): number | null => {
      for (const label of labels) {
        if (label in rawCounts) {
          return parseCount(rawCounts[label]);
        }
      }
      return null;
    };

    const avatar = info.querySelector(selectors.channelAvatar);
    const counts: ChannelCounts = {
      subscribers: countOf('subscribers', 'subscriber'),
      photos: countOf('photos', 'photo'),
      videos: countOf('videos', 'video'),
      files: countOf('files', 'file'),
      links: countOf('links', 'link'),
      raw: rawCounts,
    };

    return {
      username: handle,
      url: `https://t.me/${handle}`,
      title: textOf(info, selectors.channelTitle) || handle,
      description: textOf(info, selectors.channelDescription),
      descriptionHtml: htmlOf(info, selectors.channelDescription),
      avatarUrl: avatar?.getAttribute('src') || null,
      counts,
    };
  }

  // Try profile card (private channel fallback)
  const title = textOf(tree, selectors.cardTitle);
  if (title) {
    const avatar = tree.querySelector(selectors.cardAvatar);
    const rawCounts: Record<string, string> = {};
    let subscribers: number | null = null;

    const extraText = textOf(tree, selectors.cardExtra);
    if (extraText && extraText.includes('subscriber')) {
      rawCounts.subscribers = extraText;
      subscribers = parseCount(extraText);
    }

    return {
      username,
      url: `https://t.me/${username}`,
      title,
      description: textOf(tree, selectors.cardDescription),
      descriptionHtml: htmlOf(tree, selectors.cardDescription),
      avatarUrl: avatar?.getAttribute('src') || null,
      counts: {
        subscribers,
        photos: null,
        videos: null,
        files: null,
        links: null,
        raw: rawCounts,
      },
    };
  }

  return null;
}
